Remove race conditions from File.Move and Directory.Move

joshudson opened the pull request:
(Directory.Move only applies on Linux)

Fixes #31228

joshudson committed:
#31228 remove race conditions from File.Move and Directory.Move (Directory.Move only applies on Linux)

joshudson commented on src/libraries/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -162,56 +115,35 @@ public static void MoveFile(string sourceFullPath, string destFullPath, bool ove
-            // where hard links are not supported (such as FAT). Therefore, given that source file exists,
-            // rename is used in 2 cases: when dest file does not exist or when source path and dest
-            // path refer to the same file (on the same device). This is important for case-insensitive
+            // link/unlink are used instead. rename is used when the source path and dest path refer
```
comment about performance of trying link() on FAT is clearly not correct because the syscall fast-fails so if you have to perform any checks before deciding to rename() it's slower.

jhudsoncedaron commented:
Azure builds die with

🍺  /usr/local/Cellar/pkg-config/0.29.2: 11 files, 626.9KB
Error: python 3.7.6_1 is already installed
To upgrade to 3.7.7, run `brew upgrade python`.

##[error]Bash exited with code '1'.
Finishing: Install native dependencies

Not mine!

joshudson committed:
Merge https://github.com/dotnet/runtime

joshudson committed:
Use check_symbol_exists as well because android's stdio.h is broken

joshudson commented:
(NETCORE_ENGINEERING_TELEMETRY=InitializeToolset) Failed to install dotnet SDK from public location (exit code '1').

Not mine again!

joshudson commented:
Ok somebody else needs to look at this. Remaining failed test cases:

System.IO.Tests.DirectoryInfo_MoveTo.MoveDirectory_NoOpWhenMovingDirectoryWithUpperCaseToOtherDirectoryWithLowerCase
 System.IO.IOException : Cannot create '/var/folders/j1/76_p663903g5g2rg7ckqbv440000gy/T/DirectoryInfo_MoveTo_xz4to5q4.r0u/bar/foo' because a file or directory with the same name already exists.
   at System.IO.FileSystem.MoveDirectory(String sourceFullPath, String destFullPath) in /_/src/libraries/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:line 317
   at System.IO.Directory.Move(String sourceDirName, String destDirName) in /_/src/libraries/System.IO.FileSystem/src/System/IO/Directory.cs:line 300
   at System.IO.Tests.Directory_Move.MoveDirectory_NoOpWhenMovingDirectoryWithUpperCaseToOtherDirectoryWithLowerCase() in /_/src/libraries/System.IO.FileSystem/tests/Directory/Move.cs:line 341

This one is testing for
    Directory.Create("foo");
    Directory.Create("bar/FOO");
    Directory.Move("foo", "bar/foo");

actually succeeding and clobbering bar/FOO. I'm convinced my patch behaves correctly and the test is just wrong, because the documentation says it won't clobber the directory.

The remaining two failing tests are:

System.IO.Tests.File_Move_Tests.Unix_File_Move_To_Same_Director
 Renamed event did not occur as expected\nExpected: True\nActual:   False
   at System.IO.Tests.FileSystemWatcherTest.ExecuteAndVerifyEvents(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Boolean assertExpected, String[] expectedPaths, Int32 timeout) in /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs:line 339
   at System.IO.Tests.FileSystemWatcherTest.ExpectEvent(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Action cleanup, String[] expectedPaths, Int32 attempts, Int32 timeout) in /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs:line 227
   at System.IO.Tests.FileSystemWatcherTest.ExpectEvent(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Action cleanup, String expectedPath, Int32 attempts, Int32 timeout) in /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs:line 197
   at System.IO.Tests.File_Move_Tests.FileMove_SameDirectory(WatcherChangeTypes eventType) in /_/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Move.cs:line 142
   at System.IO.Tests.File_Move_Tests.Unix_File_Move_To_Same_Directory() in /_/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Move.cs:line 24

System.IO.Tests.File_Move_Tests.Unix_File_Move_With_Set_NotifyFilte
 Renamed event did not occur as expected\nExpected: True\nActual:   False
   at System.IO.Tests.FileSystemWatcherTest.ExecuteAndVerifyEvents(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Boolean assertExpected, String[] expectedPaths, Int32 timeout) in /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs:line 339
   at System.IO.Tests.FileSystemWatcherTest.ExpectEvent(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Action cleanup, String[] expectedPaths, Int32 attempts, Int32 timeout) in /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs:line 227
   at System.IO.Tests.FileSystemWatcherTest.ExpectEvent(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Action cleanup, String expectedPath, Int32 attempts, Int32 timeout) in /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs:line 197
   at System.IO.Tests.File_Move_Tests.FileMove_WithNotifyFilter(WatcherChangeTypes eventType) in /_/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Move.cs:line 311
   at System.IO.Tests.File_Move_Tests.Unix_File_Move_With_Set_NotifyFilter() in /_/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Move.cs:line 120

which check if File.Move() actually does a rename operation. But you can't do `rename()` and have race-free clobber avoidance here. The OS for which this is failing is Mac OSX. The only race-free solution involves `link()` and `unlink()`.

carlossanlop commented on src/libraries/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -162,57 +115,35 @@ public static void MoveFile(string sourceFullPath, string destFullPath, bool ove
-            if (overwrite)
+            // Note that the test is actually in native code, and the only thing managed code does
+            // is check fo EXDEV and call CopyFile.
+            if (Interop.Sys.Rename(sourceFullPath, destFullPath, overwrite ? 1 : 0) < 0)
```
The 0 value for the flags is not defined in the comments above the P/Invoke method Interop.Sys.Rename. Consider mentioning what it means in the comments.

joshudson replied:
Can do. It's a flags enum. I'll think of good wording soon.

joshudson replied:
I think I improved the comment enough. If not, I'll have to start commenting each call site.

joshudson committed:
Improve doc comment for Interop.Sys.Rename

adamsitnik reviewed (commented):
Hi @joshudson

First of all, I want to thank you for your work and apologize for the delay in the reviewing process.

Could you please respond to my comments and also solve the conflicts?

Thanks,
Adam

joshudson commented:
Hey just to let you know I'm not ignoring this. I've got some time off upcoming and I intend to look at this during that time.

joshudson committed:
Merge remote-tracking branch 'upstream/master'

Jozkee commented on src/libraries/Native/Unix/System.Native/pal_io.c:
```diff
@@ -995,10 +1001,114 @@ int32_t SystemNative_ReadLink(const char* path, char* buffer, int32_t bufferSize
+#if HAVE_RENAMEAT2 && defined(RENAME_NOREPLACE)
+        // Linux renameat2() provides the functionality we want inline.
+        // This is in a feature check block as this API may well be ported to other OSes later.
+        do {
```
nit
```suggestion
        do 
        {
```

Jozkee commented on src/libraries/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -129,7 +82,7 @@ public static void ReplaceFile(string sourceFullPath, string destFullPath, strin
+            Interop.CheckIo(Interop.Sys.Rename(sourceFullPath, destFullPath, 1));
         }
 
         public static void MoveFile(string sourceFullPath, string destFullPath)
```
Is the TICTOU bug also possible to reproduce on `File.Move` and `File.Replace`? I am trying to understand why these methods were also modified.

joshudson replied:
Yes. See PR title.

joshudson commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.Rename.cs:
```diff
@@ -14,10 +14,12 @@ internal static partial class Sys
         /// </summary>
         /// <param name="oldPath">Path to the source item</param>
         /// <param name="newPath">Path to the desired new item</param>
+        /// <param name="flags">bit flags: flags &amp; 1: whether to overwrite target or not; flags &amp; 2: whether this is Directory.Move or not</param>
```
I may or may not be able to get enum working correctly. If it doesn't work, I can use `const` though.

joshudson commented on src/libraries/Native/Unix/System.Native/pal_io.c:
```diff
@@ -1100,10 +1106,114 @@ int32_t SystemNative_ReadLink(const char* path, char* buffer, int32_t bufferSize
+    }
+    else
+    {
+#if HAVE_RENAMEAT2 && defined(RENAME_NOREPLACE)
```
It's feature probing. Clearly both must exist or it won't work. Officially, we don't need to know. I think it's possible, but I'm not certain.

joshudson commented on src/libraries/Native/Unix/System.Native/pal_io.c:
```diff
@@ -1100,10 +1106,114 @@ int32_t SystemNative_ReadLink(const char* path, char* buffer, int32_t bufferSize
+            else if (errno == EEXIST)
+            {
+                // But renameat2 returns EEXIST on renaming a file onto itself. Check that.
+                result = RenameOnEexist(result, NULL, oldPath, newPath);
```
No unit test is possible because we can't demand root to run unit tests. We can't get here if the name components are equal.

joshudson commented on src/libraries/Native/Unix/System.Native/pal_io.c:
```diff
@@ -1100,10 +1106,114 @@ int32_t SystemNative_ReadLink(const char* path, char* buffer, int32_t bufferSize
+        // Linux renameat2() provides the functionality we want inline.
+        // This is in a feature check block as this API may well be ported to other OSes later.
+        do {
+            result = renameat2(AT_FDCWD, oldPath, AT_FDCWD, newPath, RENAME_NOREPLACE);
```
I'm calling renameat2() for RENAME_NOREPLACE. AT_FDCWD is what you pass if you don't have anything else to pass in that slot.

joshudson replied:
On third attempt, pronoun [it] is ambiguous. Do you mean AT_FDCWD or RENAME_NOREPLACE? A unit test for missing RENAME_NOREPLACE would just check that MoveFile(,,false) doesn't ovewrite its target and I think that already exists. If AT_FDCWD isn't passed, it will fail pretty hard as 0 won't be a directory.

joshudson commented on src/libraries/Native/Unix/System.Native/pal_io.c:
```diff
@@ -1100,10 +1106,114 @@ int32_t SystemNative_ReadLink(const char* path, char* buffer, int32_t bufferSize
+                result = link(oldPath, newPath);
+                if (result >= 0)
+                {
+                    while ((result = unlink(oldPath)) < 0 && errno == EINTR);
```
Side effect: File system watcher looking for rename() won't ever see it.

Incidentally I had to go through our own codebase ten years ago and systematically swap File.Move() for P/Invoke of CreateHardLink because somebody deployed our application on Windows pointing at a NAS box and hit the race condition where Windows resolving a UNC to a NAS box will inherently have this race. There is one other side effect, that is if the application crashes, you can see both names at once. But it's not really a side effect on Unix, because *that can happen anyway* on power loss.

joshudson commented on src/libraries/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -37,82 +37,35 @@ public static void Decrypt(string path)
-                // file, if it exists.  If deletion fails for a reason other than the file not existing, fail.
-                if (Interop.Sys.Unlink(destBackupFullPath) != 0)
+                // We're backing up the destination file to the backup file. If it exists, it will be removed if possible.
+                if (Interop.Sys.Rename(destFullPath, destBackupFullPath, 1) < 0)
```
It's supposed to be.

joshudson commented on src/libraries/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -37,82 +37,35 @@ public static void Decrypt(string path)
-                // file, if it exists.  If deletion fails for a reason other than the file not existing, fail.
-                if (Interop.Sys.Unlink(destBackupFullPath) != 0)
+                // We're backing up the destination file to the backup file. If it exists, it will be removed if possible.
+                if (Interop.Sys.Rename(destFullPath, destBackupFullPath, 1) < 0)
```
That's a style debate, not a functional debate. I kept the original return values of 0 for success and -1 for failure.

joshudson commented on src/libraries/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -37,82 +37,35 @@ public static void Decrypt(string path)
+                        errno.Error == Interop.Error.EMLINK)       // too many hard links to the source file
+                   {
+                        // Destination file could not be renamed. Copy it.
+                        CopyFile(destFullPath, destBackupFullPath, true);
```
Because EMLINK = too many links to file aready. Says so right in the comment above too.

joshudson commented on src/libraries/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -37,82 +37,35 @@ public static void Decrypt(string path)
+                   {
+                        // Destination file could not be renamed. Copy it.
+                        CopyFile(destFullPath, destBackupFullPath, true);
+                        if (Interop.Sys.Unlink(destBackupFullPath) != 0)
```
I'll look into it once I can run tests again.

joshudson replied:
The call to CopyFile remains in case rename didn't work because it was cross-filesystem. Basically, the if does what LinkOrCopyFile() used to do.

joshudson commented:
I'm blocked on https://github.com/dotnet/runtime/issues/46301 ; the tree doesn't build referencing SOS_readme.md

joshudson committed:
Fix minor issues from review

joshudson commented:
I fixed minor problems from review; however I ran into problems dealing with File.Replace. Filed doc issue https://github.com/dotnet/docs/issues/22288 ; the code might also be wrong in master.

joshudson committed:
Revert changes to ReplaceFile() as the documentation is not reliable

joshudson committed:
Merge https://github.com/dotnet/runtime

danmosemsft commented:
Some test failures:
```
    System.IO.Tests.File_Move_Tests.Unix_File_Move_With_Set_NotifyFilter [FAIL]
      Renamed event did not occur as expected
      Expected: True
      Actual:   False
      Stack Trace:
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs(337,0): at System.IO.Tests.FileSystemWatcherTest.ExecuteAndVerifyEvents(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Boolean assertExpected, String[] expectedPaths, Int32 timeout)
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs(225,0): at System.IO.Tests.FileSystemWatcherTest.ExpectEvent(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Action cleanup, String[] expectedPaths, Int32 attempts, Int32 timeout)
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs(195,0): at System.IO.Tests.FileSystemWatcherTest.ExpectEvent(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Action cleanup, String expectedPath, Int32 attempts, Int32 timeout)
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Move.cs(311,0): at System.IO.Tests.File_Move_Tests.FileMove_WithNotifyFilter(WatcherChangeTypes eventType)
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Move.cs(120,0): at System.IO.Tests.File_Move_Tests.Unix_File_Move_With_Set_NotifyFilter()
      Renamed event did not occur as expected
      Expected: True
      Actual:   False
    System.IO.Tests.File_Move_Tests.Unix_File_Move_In_Nested_Directory(includeSubdirectories: True) [FAIL]
      Stack Trace:
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs(337,0): at System.IO.Tests.FileSystemWatcherTest.ExecuteAndVerifyEvents(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Boolean assertExpected, String[] expectedPaths, Int32 timeout)
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs(225,0): at System.IO.Tests.FileSystemWatcherTest.ExpectEvent(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Action cleanup, String[] expectedPaths, Int32 attempts, Int32 timeout)
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs(195,0): at System.IO.Tests.FileSystemWatcherTest.ExpectEvent(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Action cleanup, String expectedPath, Int32 attempts, Int32 timeout)
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Move.cs(290,0): at System.IO.Tests.File_Move_Tests.FileMove_NestedDirectory(WatcherChangeTypes eventType, Boolean includeSubdirectories)
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Move.cs(106,0): at System.IO.Tests.File_Move_Tests.Unix_File_Move_In_Nested_Directory(Boolean includeSubdirectories)
    System.IO.Tests.File_Move_Tests.Unix_File_Move_To_Same_Directory [FAIL]
      Renamed event did not occur as expected
      Expected: True
      Actual:   False
      Stack Trace:
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs(337,0): at System.IO.Tests.FileSystemWatcherTest.ExecuteAndVerifyEvents(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Boolean assertExpected, String[] expectedPaths, Int32 timeout)
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs(225,0): at System.IO.Tests.FileSystemWatcherTest.ExpectEvent(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Action cleanup, String[] expectedPaths, Int32 attempts, Int32 timeout)
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs(195,0): at System.IO.Tests.FileSystemWatcherTest.ExpectEvent(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Action cleanup, String expectedPath, Int32 attempts, Int32 timeout)
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Move.cs(142,0): at System.IO.Tests.File_Move_Tests.FileMove_SameDirectory(WatcherChangeTypes eventType)
        /_/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Move.cs(24,0): at System.IO.Tests.File_Move_Tests.Unix_File_Move_To_Same_Directory()
```

joshudson commented:
@danmosemsft : Yes, that's a decision that has to be made. I left these failing until we decide we're good with the filesystem watcher not seeing any rename events on `File.Move(..., ..., false);` until the build target has a high enough libc, or carry around the syscall definition of `renameat2()` so we can call it w/o libc.

danmosemsft commented:
@adamsitnik thoughts about next steps for this PR?

danmoseley commented:
We don't seem to be making progress on this PR. Regretfully, I'm going to close it. It can always be reopened .
